pkg/mutation/system.go

```diff
@@ -1,20 +1,173 @@
 package mutation
 
 import (
+	"fmt"
+	"sort"
 	"sync"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
 // System keeps the list of mutations and
 // provides an interface to apply mutations.
 type System struct {
-	// schemaDB schema.DB commented to please the linter, will uncomment in the implementation
-	mutators []Mutator
+	schemaDB        SchemaDB
+	orderedMutators []Mutator
+	mutatorsMap     map[ID]Mutator
 	sync.RWMutex
 }
 
 // NewSystem initializes an empty mutation system
 func NewSystem() *System {
 	return &System{
-		mutators: make([]Mutator, 0),
+		orderedMutators: make([]Mutator, 0),
+		mutatorsMap:     make(map[ID]Mutator),
+	}
+}
+
+// Upsert updates or insert the given object, and returns
+// an error in case of conflicts
+func (s *System) Upsert(m Mutator) error {
+	s.Lock()
+	defer s.Unlock()
+
+	current, ok := s.mutatorsMap[m.ID()]
+	if ok && !m.HasDiff(current) {
+		return nil
+	}
+
+	toAdd := m.DeepCopy()
+
+	// Checking schema consistence only if the mutator has schema
+	if withSchema, ok := toAdd.(MutatorWithSchema); ok {
+		err := s.schemaDB.Upsert(withSchema)
+		if err != nil {
+			return errors.Wrapf(err, "Schema upsert failed")
+		}
+	}
+
+	s.mutatorsMap[toAdd.ID()] = toAdd
+
+	i := sort.Search(len(s.orderedMutators), func(i int) bool {
+		return greaterOrEqual(s.orderedMutators[i], toAdd)
+	})
+
+	if i == len(s.orderedMutators) { // Adding to the bottom of the list
+		s.orderedMutators = append(s.orderedMutators, toAdd)
+		return nil
+	}
+
+	found := equal(s.orderedMutators[i], toAdd)
+	if found {
+		s.orderedMutators[i] = toAdd
+		return nil
+	}
+
+	s.orderedMutators = append(s.orderedMutators, nil)
+	copy(s.orderedMutators[i+1:], s.orderedMutators[i:])
+	s.orderedMutators[i] = toAdd
+	return nil
+}
+
+// Mutate applies the mutation in place to the given object
+func (s *System) Mutate(obj *unstructured.Unstructured, ns *corev1.Namespace) error {
+	s.RLock()
+	defer s.RUnlock()
+
+	maxIterations := len(s.orderedMutators) + 1
+
+	old := obj.DeepCopy()
+
+	for i := 0; i < maxIterations; i++ {
+		for _, m := range s.orderedMutators {
+			if m.Matches(obj, ns) {
+				err := m.Mutate(obj)
+				if err != nil {
+					return errors.Wrapf(err, "Mutation failed for %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind)
+				}
+			}
+		}
+		if cmp.Equal(old, obj) {
+			return nil
+		}
+		old = obj.DeepCopy()
+	}
+
+	return fmt.Errorf("Mutation not converging for %s %s", obj.GroupVersionKind().Group, obj.GroupVersionKind().Kind)
+}
+
+// Remove removes the mutator from the mutation system
+func (s *System) Remove(m Mutator) error {
+	s.Lock()
+	defer s.Unlock()
+
+	if _, ok := s.mutatorsMap[m.ID()]; !ok {
+		return nil
+	}
+
+	err := s.schemaDB.Remove(m.ID())
+	if err != nil {
+		return errors.Wrapf(err, "Schema remove failed")
```

Comment:
I'm not sure how we're recover from a failed remove ... we'd probably have to rebuild schemaDB from scratch.

We should see if it's possible for Remove() to return an error status before worrying too much about this.

Reply:
I guess it depends on the reason why remove would fail.

We already discussed [here](https://github.com/open-policy-agent/gatekeeper/pull/928#discussion_r515437537) that remove should not return an error if the object does not exist, which leaves out worse scenarios.
Depending on the schemaDB implementation we may even find that we don't need `Remove` to return an error at all, or that the "db" is corrupted (and in that case we can try to rebuild it and scream if the rebuid fails).

Reply:
+1 to the seeing if there is a reason for an implementation to return error at all. I think we're in the "not enough information" stage of how to handle this. Maybe leave a TODO here so we can re-evaluate once all the pieces are available?

Reply:
Will do